roles/hmsdocker/templates/docker-compose.yml.j2

```diff
@@ -434,8 +433,9 @@ services:
     {% elif hmsdocker_transmission_custom_download_path_enabled %}
       - {{ transmission_custom_download_path }}/torrents:/data/torrents
     {% else %}
-      - ${HMSD_APPS_PATH}/transmission/torrents:/data/torrents
+      - {{ hms_docker_mount_path }}:/data
```

Comment:
Should this also be `${HMSD_DOWNLOAD_PATH}`?

Reply:
the way i always set mine up is with everything mount to /Media Server and inside app you would set your path to the appropriate folder.  This may be a handy reference for you as far as paths and links

https://trash-guides.info/Hardlinks/Hardlinks-and-Instant-Moves/#hardlinks-limitations

Reply:
So should all apps have `${HMSD_MOUNT_PATH}` mounted to the container `/data` directory then?

Reply:
> So should all apps have `${HMSD_MOUNT_PATH}` mounted to the container `/data` directory then?
 The directory name doesn't necessarily matter as long as it is the same for every one.  I have about 15 set up on my new docker server on truenas and every app has this path

    volumes:
      - /mnt/docker/data/plex/config:/config
      - /mnt/Media Server:**/Media Server**

As long as every app points to the same path, the bolded text, it can be named whatever you want.

So it could be /data, /dogs, /cats.....  anything you want.

So how mine works, 
Plex  /Media Server/Movies
        /Media Server/TV Shows
Sonarr
       same as above
       /Media Server/Torrents/Completed
Qbittorrent
       /Media Server/Torrents/Downloading
       /Media Server/Torrents/Completed
       /Media Server/Torrents/Downloading Torrent Files
       /Media Server/Torrents/Completed Torrent Files

And so on, but since every one of them follows the same path it allows for atomic moves and hard linking to work properly

Reply:
Do atomic moves and hard linking work with NFS and CIFS network shares?

Reply:
yes, as long as its in a single mount.  At least i confirmed it with my NFS on my last setup

So this will work
 /server:/mnt/server
then in both app containers point to /mnt/server/torrents/download and /mnt/server/movies

but this will not

/server/torrent/download:/download
/server/movies:/movies

when you have individual mounts like the second example which is what ALOT of people think is easiest/best the big problem is that it sees it as 2 diff drives so even though it super simplifies your structure cause its all on root and clearly labeled it cannot used advanced features that are in place for same drive moves because as far as its concered they are not a single drive

Reply:
Hi guys, I'm little "behind" here. I just see that you are both on the good track.
What is essential to understand is as i2sly mentioned, for hardlinking to work the files needs to be on the same filesystem.

In a Linux filesystem, a hard link is a reference to a file's inode, allowing multiple filenames to point to the same data on disk. Unlike symbolic links, hard links are indistinguishable from the original file; they share the same inode number. Changes made to the file via any hard link reflect across all linked names. Deleting a hard link reduces the inode's link/referral count (refcount), and the file is only removed when all hard links are deleted, (when the refcount is reduced to zero). Hard links can't span different filesystems and can't be created for directories to avoid complications. 
Although they may well be on the host /opt/hms-docker, if you mount host subfolders inside of different mount-points within the containers, from the container perspective the two mount-points are different filesystems. Hardlinking works with NFSv3 and NFSv4 on remote Linux FS, but not with NFSv2 (almost non existant nowadays).

On Windows NTFS, it is also possible to create hardlinks with :
mklink /H link_name target_file
But because of the current limitations of the CIFS/SMB protocol, hardlinks will be not created on remote NTFS filesystems if remotely mounted by CIFS.

Reply:
> Hi guys, I'm little "behind" here. I just see that you are both on the good track. What is essential to understand is as i2sly mentioned, for hardlinking to work the files needs to be on the same filesystem.
> 
> In a Linux filesystem, a hard link is a reference to a file's inode, allowing multiple filenames to point to the same data on disk. Unlike symbolic links, hard links are indistinguishable from the original file; they share the same inode number. Changes made to the file via any hard link reflect across all linked names. Deleting a hard link reduces the inode's link/referral count (refcount), and the file is only removed when all hard links are deleted, (when the refcount is reduced to zero). Hard links can't span different filesystems and can't be created for directories to avoid complications. Although on the host /opt/hms-docker is contained in a single filesystem, if you mount host subfolders inside of different mount-points within the containers, from the container perspective the separate mount-points are different filesystems which prevents any hardlinking to be performed from one to the other. Hardlinking works with NFSv3 and NFSv4 on remote Linux FS, but not with NFSv2 (almost non existant nowadays).
> 
> On Windows NTFS, it is also possible to create hardlinks with : mklink /H link_name target_file But because of the current limitations of the CIFS/SMB protocol, hardlinks will be not created on remote NTFS filesystems if remotely mounted by CIFS.

Very well put, thats why i gave him the link to the guide on linking and screenshots of my structure use hoping to clarify it as much as possible.  I have almost got my migration from the kubernetics backend done, all i gotta do is get my traefik and gluetun setup.  Once I get it all done I will see about giving myself a crash course in ansible and create a new jail to test out your setup and see if we can make some magic happen.

I am currently running Truenas Scale on barebones.  I just switched from there kuber apps to running a jailmaker sandbox so set it all up by hand so my production server was solid.  Once I get a new sandbox setup I will see if this works under sandbox and if not I will have to install a VM to test.

Reply:
Also Ahembree,  just thought about it and maybe you have had this happen but not really known why but other than the obvious space saving of hard linking or the speed of atomic moves for me the number one reason for doing this is system resource management.  My server is 100tb, 128gb ram, i9 12900k processor with my main nas being on 8 22tb reds in zfs.  With that being said, if I had large torrents as in 100gb+ or multiple 20+gb torrents end all around the same time the I/O would cause a strain on the system that would often cause the qbittorrent, transmission and rutorrent alike to crash and reload.  That being a minor inconvenience had me accepting that it was just part of doing shit at this level but a year ago I actually reconfigured my server from the ground up and did it MY way instead of following the way just about every install guide suggests and incedentally I realized looking at the logs a few months in that I had no seemingly random crashes in my torrent apps and also eliminated all heat spikes in my hard drive logs since I was no longer over taxing them.  So that is a HUGE reason for proper linking and honestly suprises me that more guides don't set up there servers properly.  Was looking at one that looks very well done, Deployarr, and even he still uses /downloads mount point so hardlinking/atomic won't work on that one either

Reply:
> end all around the same time the I/O would cause a strain on the system that would often cause the qbittorrent, transmission and rutorrent alike to crash and reload

I always attributed this to high load but never really thought about digging into it further, thanks for this!

Reply:
Just added versioning into the `master` branch, so I'd like to merge those changes into this PR before merging this PR so that versioning is correct

Reply:
If it's possible to standardize on one single volume mount for all media/downloads across all containers that need it, such as `${HMSD_MOUNT_PATH}:/data`, I'd prefer to go that route to reduce complexity.

And then how could this work for when there are separate instances of Sonarr and Radarr for 4K? How would this directory mapping work and would it support hard links as well? Or should I just remove a separate 4K media mount entirely?

I currently have a separate NFS share for 4K content but if that's bad practice and wouldn't support hard links, I can easily switch it

Reply:
> If it's possible to standardize on one single volume mount for all media/downloads across all containers that need it, such as `${HMSD_MOUNT_PATH}:/data`, I'd prefer to go that route to reduce complexity.
> 
> And then how could this work for when there are separate instances of Sonarr and Radarr for 4K? How would this directory mapping work and would it support hard links as well? Or should I just remove a separate 4K media mount entirely?
> 
> I currently have a separate NFS share for 4K content but if that's bad practice and wouldn't support hard links, I can easily switch it

As I have said before, one path for all is the way to go.  So for your examples stated  it would go as follows

${HMSD_MOUNT_PATH}:/data

in Sonarr path is /data/TV Shows
in Radarr path is /data/Movies
in 4k Sonarr /data/4K TV Shows
in 4K Radarr /data/4K Movies
in download app /data/Torrents/Downloading
                            /data/Torrents/Completed
in unpackarr /data/Torrents/Completed

Reply:
I've got the following in my local code and I'm curious if it works with hardlinks.

In order to keep supporting the additional NAS mounts (if people have another, https://github.com/ahembree/ansible-hms-docker/issues/61), I've mounted the library to `/data/<library name>` so that `/data/<library name>_additional` could potentially also support hardlinks, which is what I'm curious about in order to support hardlinks for the additional NAS mounts.

Mounts are as follows:
```
# Sabnzbd
- ${HMSD_DOWNLOAD_PATH}/usenet/sabnzbd:/data

# Nzbget
- ${HMSD_DOWNLOAD_PATH}/usenet/nzbget:/data

# Transmission
# This is kept as HMSD_MOUNT_PATH due to the environment variables defining the rest of the path that equates to HMSD_DOWNLOAD_PATH
- ${HMSD_MOUNT_PATH}:/data
...
# Transmission env vars:
- TRANSMISSION_DOWNLOAD_DIR=/data/{{ hms_docker_downloads_folder_name }}/torrents/transmission/
- TRANSMISSION_INCOMPLETE_DIR=/data/{{ hms_docker_downloads_folder_name }}/torrents/transmission/incomplete
- TRANSMISSION_WATCH_DIR=/data/{{ hms_docker_downloads_folder_name }}/torrents/transmission/watch

# Sonarr and Radarr (including 4K's) and Readarr
- ${HMSD_LIBRARY_PATH}:/data/{{ hms_docker_library_folder_name }}
- ${HMSD_DOWNLOAD_PATH}:/data/{{ hms_docker_downloads_folder_name }}
- <for loop to go over other mount paths>:/data/{{ hms_docker_downloads_folder_name }}_additional/<for loop to go over names>

# Plex / Jellyfin / Emby / Tdarr / Kavita / Calibre
- ${HMSD_LIBRARY_PATH}:/data/{{ hms_docker_library_folder_name }}
- <for loop to go over other mount paths>:/data/{{ hms_docker_downloads_folder_name }}_additional/<for loop to go over names>
```

As much as I want to reduce complexity by reducing variables and paths, I will need to support the capability to add additional NAS mounts. Will this work?

Reply:
> I've got the following in my local code and I'm curious if it works with hardlinks.
> 
> In order to keep supporting the additional NAS mounts (if people have another, #61), I've mounted the main mount path to `/data/media` so that `/data/media_additional` could potentially also support hardlinks, which is what I'm curious about in order to support hardlinks for the additional NAS mounts.
> 
> Mounts are as follows:
> 
> ```
> # Sabnzbd
> - ${HMSD_DOWNLOAD_PATH}/usenet/sabnzbd:/data/media
> 
> # Nzbget
> - ${HMSD_DOWNLOAD_PATH}/usenet/nzbget:/data/media
> 
> # Transmission
> # This is kept as HMSD_MOUNT_PATH due to the environment variables defining the rest of the path that equates to HMSD_DOWNLOAD_PATH
> - ${HMSD_MOUNT_PATH}:/data/media
> ...
> # Transmission env vars:
> - TRANSMISSION_DOWNLOAD_DIR=/data/media/{{ hms_docker_downloads_folder_name }}/torrents/transmission/
> - TRANSMISSION_INCOMPLETE_DIR=/data/media/{{ hms_docker_downloads_folder_name }}/torrents/transmission/incomplete
> - TRANSMISSION_WATCH_DIR=/data/media/{{ hms_docker_downloads_folder_name }}/torrents/transmission/watch
> 
> # Sonarr and Radarr (including 4K's) and Readarr
> - ${HMSD_MOUNT_PATH}:/data/media
> - <for loop to go over other mount paths>:/data/media_additional/<for loop to go over names>
> 
> # Plex / Jellyfin / Emby / Tdarr / Kavita / Calibre
> - ${HMSD_LIBRARY_PATH}:/data/media
> - <for loop to go over other mount paths>:/data/media_additional/<for loop to go over names>
> ```
> 
> As much as I want to reduce complexity by reducing variables and paths, I will need to support the capability to add additional NAS mounts. Will this work?
> 
> Update: after realizing I should just get a shell in a container, I now see why this won't work. Will both hardlinks and normal copies (to support the additional NAS paths) continue to work correctly?

You are correct in realizing this will not work.  Your comment about adding a new mount and it working will never work for hard linking.  It has to be the same drive.  So not only does the path apps see have to have identical paths but it ALSO does have to indeed be on the same path.  So you can add a dozen mounts with no problem but everything to do with downloading and media data must be on one path.  So as i have demonstrated a few times  lets say your NTFS Share you should have your movies/tv/torrents/sub all in one folder on the sharing pc.  So they can be at the root or they can be at /Server or Server/Stuff/Arrs as long as that data source points to the same /mounted_folder on every app

So from every app it sees your base folder as mounted_folder.   if you have radarr mounted_folder/radarr mount to /radarr and sonarr mounted_folder/sonarr mount to /sonarr it sees them as distintly different.

it would have to be mounted_folder and each app sees mounted_folder/sonarr and mounted_folder/radarr and mounted_folder/torrent/completed

as long as each starts at mounted_folder it will successfully hardlink

if you would like to go over it in a more fluid manner you can add me on discord with user i2sly or on telegram at @jhuston82

Reply:
Sent request on Discord, same username as GitHub.

I want to ensure there is still support for these additional NAS mounts, even if there is no hardlink support for that specific mount (assuming the Arr apps fall back to copy if hardlink fails).